### ISSUE TITLE
Our Technology Stack

### ISSUE THREAD

**Issue by marcotroisi** [easy-author/vox#2]
# Server Side

We want everyone to be able to install and set up **Vox**, so the best choice seems to be pretty uncontroversial here:
- PHP
# Server Side Framework
- Moss
- Fat Free
- Slim
- PHP 4 :D 
# ORM
- Eloquent
- Doctrine 
# Client Side framework/language
- Javascript with Angularjs
- Javascript with jQuery
- Dart
# Theming
## Server
- Pure PHP
- phpQuery
- Twig
## Client
- Angular or jQuery, depending on what we choose as a client side framework
# Editor
- Markdown (with preview)
- Wysiwyg
- Both

**Comment by marcotroisi** [comment]
My personal choices/opinions:
- **Server side framework**: Fat Free _or_ Moss
- **ORM**: Doctrine
- **Client side framework**: Javascript with Angularjs
- **Theming, Server**: phpQuery or Twig
- **Theming, Client**: Angularjs
- **Editor**: Markdown with live preview and shortcuts

**Comment by gerty3000** [comment]
Hi :)
if I can have something to say I choose:
- **Server side framework**: Slim or Moss
- **ORM**: Doctrine
- **Client side framework**: Dart, because imho looks friendly for programmers
- **Theming, Server**: Twig (I guess phpQuery code is terrible)
- **Theming, Client**: -
- **Editor**: Markdown

**Comment by potfur** [comment]
My vision:
- **Server side framework**: Slim or Moss
- **ORM**: none (data layer should be covered with defined repository interfaces, this allows for different implementations with different ORM or plain SQL to increase speed, or even switch to files)
- **Client side framework**: -
- **Theming, Server**: plain php behind interface (same as in ORM, we can add other later)
- **Theming, Client**: -
- **Editor**: Markdown

\- means that we should leave as much freedom as we can

**Comment by marcotroisi** [comment]
@seclu thanks for joining the decisional process, your opinion is always appreciated - not sure about Dart, it may just be an unneeded overhead for us developers. It looks very good, but how many developers can actually work with it?

@potfur on the ORM front, do you mean that we can enable our favourite ORM for each Database? Would you explain? :) 

**Update:** To clarify: the _client side framework_ is the one that **we** are going to use to design the interface of the admin panel, so I think we should at least choose one, to make our development easier.

**Comment by potfur** [comment]
@marcotroisi we can create let say BlogPostRepositoryInterface that defines methods for accessing conrete entry or list of entries for third page or saving/updating them. Then we can create classes implementing this interface for whatever tool we wan - for ORM, Mongo, plain SQL or even store everything in files.

**Comment by marcotroisi** [comment]
@potfur I see what you are saying. But wouldn't it be almost the same thing to use an ORM that already support as many Databases as possible?

**Comment by gerty3000** [comment]
After thought I guess @potfur has right. We should have an interface with appropriate methods as boundary for model layer. But I still think that ORM is good decision. I know more or less Doctrine, but  maybe we should look for example Propel?

**Comment by potfur** [comment]
@marcotroisi this _almost_ is rather big difference. We don't need to decide and limit ourselfs to one ORM/technology. Behind those interfaces we still can use ORM, but at any point we can implement new classes for something else.

**update** for backend we should use bootstrap or something similar.

**Comment by marcotroisi** [comment]
@potfur you have a valid point in saying that we should make it as "abstract" as possible. I agree wholeheartedly. Nevertheless I think it's fair to say that we should at least decide an ORM or, better, a database that we want to support to begin with. In a first phase, I think we should all focus on a single database, which in my opinion should be one that has high availability in the hosting world, i.e. MySql or SQLite.

**Comment by potfur** [comment]
@marcotroisi I think we can start with MySQL. And any of us can implement his own repositories for his favorite ORM

**Comment by marcotroisi** [comment]
@potfur so what you are saying is everyone can implement his favourite ORM for MySql? But then, we are going to have bits of code that use one ORM, bits that use the other Driver, etc?

**Comment by potfur** [comment]
@marcotroisi I mean that you can implement every repository using ORM and make them available in separate repository. I could do the same with other ORM or plain SQL.

**Comment by marcotroisi** [comment]
@potfur Don't you think that will cause a lot of fragmentation?

**Comment by gerty3000** [comment]
@marcotroisi you said about jQuery and Angular. For me jQuery it's convenient for DOM operation and only for this. Angular is MV(something) framework. I want to say that is different point of view about View layer behavour. If we want write any logic in this layer I choose Dart because it look quite simple. I dont know how many devs know it, but I think that is rather small number :) Angular is magic for me and I don't understand this fw.
However most of programmers have knowledge about jQuery, and for me it is a good reason to choose it.

**Comment by potfur** [comment]
@marcotroisi Yes it will. But when we choose one of the ORMs we will be bound to it.
Eloquent is active record pattern - each entity extends basic class.
Doctrine has its ArrayCollection that for 1-N and N-N relations.

Therefore at least at beginning I would stay with repository interface and plain SQL or any other ORM that is not bound to entities: [RedBean](http://redbeanphp.com/), [Moss Storage](https://github.com/potfur/moss-storage) or anything else that you know of

**Comment by gerty3000** [comment]
@potfur I think RedBean isn't the best because it looks like a proposal for creating an anemic model (correct me if I'm not right). Moss Storage is a better proposal, but... will you have enough time to develop it?

**Comment by marcotroisi** [comment]
For the ORM at least I would really like to use an established library, that we don't have to maintain ourselves but that we can really rely on. I really like RedBean, but I didn't propose it because it doesn't support Mongo and other NoSQL DBs.

**Comment by potfur** [comment]
@seclu Storage is finished.
@marcotroisi We need to decide

**Comment by marcotroisi** [comment]
@potfur my whole point about using an ORM is that it makes it a heck of a
lot easier to use different database engines - even we don't do it now. If
we start writing plain old SQL queries, which is okay for me, later on it
is going to be a hard job to port everything to SQLite or Mongo or anything
else. Why not planning well now for the future?

**Comment by potfur** [comment]
@marcotroisi Don't you think that I know that? :) I would choose Doctrine if there would be a slim possibility that we would use more than small part of its functionality. RedBean is limited to SQL databases, but for start that should not be a problem for us.

And there are active record orms like Eloquent which is no-go for me.

We can also try to use [Aura components](http://auraphp.com/packages/v2) or something else that we can find and has meaningful community.

**Comment by marcotroisi** [comment]
@potfur it is true that we add support for Mongo, for example, we are always going to change and adapt something, even if we use an ORM. 

Aura looks brilliant. I was looking for a nice query builder, to keep queries nice and readable, we should probably use theirs?

**Comment by marcotroisi** [comment]
So I would say, let's go for one between RedBean and Aura query builder and let's use it with MySQL (and SQLite) to begin with. Support for Mongo will come later on.

**Comment by potfur** [comment]
Just realized - redbean has no composer support.

**Comment by marcotroisi** [comment]
@potfur I just found this?
https://packagist.org/packages/gabordemooij/redbean

**Comment by potfur** [comment]
@marcotroisi _This Github account is for development only._ heh...
After reconsideration - lets start with doctrine, we can always change this.

**Comment by marcotroisi** [comment]
@potfur you're the best! What made you change your mind? Now I don't want to use anymore :P

**Comment by potfur** [comment]
@marcotroisi For this project, Doctrine is overkill, but - since it implements data mapper is easy to replace with something else. This allows us to postpone decision - maybe we will find something different or check other approaches.

Basically I don't care what's behind repository interfaces, as long as it stays there.
Personally I still vote for plain queries :)

**Comment by potfur** [comment]
_ping_
So, any decisions?

**Comment by marcotroisi** [comment]
"Moss Storage" is in my opinion a good middle ground. What we need at the
moment is in fact a solid, simple and extensible query builder so that we
can start easily with SQLite and MySql support out of the box.

So, to summarise, we want to use:
- Moss as a framework
- Moss Storage as a query builder and active record handler
- PHP as a server side templating language on the admin
- Twig as a server side templating language for the themes
- Angular.js as a client side templating language on the admin
- for the themes people will be able to pick their own favourite js
  framework

It seems to me that we have enough to start with. Next, we are going to
write down a roadmap for the project.
